File: src/static/js/service_map.js
/*
 * Copyright 2019 Green Valley Belgium NV
 * NOTICE: THIS FILE HAS BEEN MODIFIED BY GREEN VALLEY BELGIUM NV IN ACCORDANCE WITH THE APACHE LICENSE VERSION 2.0
 * Copyright 2018 GIG Technology NV
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @@license_version:1.6@@
 */

$(function () {
	
	var getParameterByName = function (name) {
	    var match = RegExp('[?&]' + name + '=([^&]*)').exec(window.location.search);
	    return match && decodeURIComponent(match[1].replace(/\+/g, ' '));
	};
	
	var htmlEncode = function (value) {
		value = value.replace("\r", "");
		var lines = value.split(/\n/);
		var paragraphs = [];
		$.each(lines, function (i, str) {
			paragraphs.push($('<div/>').text(str).html());
		});
		return paragraphs.join('<br>');
	};
	
	var lat_center = getParameterByName('lat');
	var lon_center = getParameterByName('lon');
	var center = lat_center && lon_center ? new google.maps.LatLng(parseFloat(lat_center), parseFloat(lon_center)) : new google.maps.LatLng(50.623211, 4.438007);
	var zoom = parseInt(getParameterByName('zoom') || '8');
	
    var mapOptions = {
    		zoom: zoom,
    		center: center
	};
    var map = new google.maps.Map(document.getElementById('map-canvas'), 
    		mapOptions);
    
    var app_id = getParameterByName("app_id") || null;
    var markers = {};
    var infoWindow = null;
    
    var last_call = new Date().getTime();
    
    var getSearchCriteria = function () {
    	last_call = new Date().getTime();
        var center = map.getCenter();
        var map_bounds = map.getBounds();
        var distance = google.maps.geometry.spherical.computeDistanceBetween(map_bounds.getNorthEast(), map_bounds.getSouthWest());
        var radius_in_km = Math.round(distance/2000,0) + 1;
        var params = {
			lat: center.lat(),
			lon: center.lng(),
			distance: radius_in_km,
			last_call: last_call
        };
        if (app_id)
        	params['app_id'] = app_id;
        return params;
    };
    
    var getServices = function (search_criteria, cursor) {
    	if (search_criteria.last_call != last_call)
    		return;
        if (cursor)
        	search_criteria['cursor'] = cursor;
        $.ajax({
        	url: '/mobi/rest/service_map/load',
        	data: search_criteria,
        	success: function (data) {
        		if (data.cursor && search_criteria.last_call == last_call)
        			getServices(search_criteria, data.cursor);
        		$.each(data.services, function (i,service) {
            		var marker = markers[service.hash];
            		if (!marker) {
            			var coords = new google.maps.LatLng(service.lat, service.lon);
            			marker = new google.maps.Marker({
            				position: coords,
                  	      	map: map,
                  	      	cursor: 'pointer'
                  	  	});
            			markers[service.hash] = marker;
            			
            			google.maps.event.addListener(marker, 'click', function() {
            				if (infoWindow)
            					infoWindow.close()
            				infoWindow = new google.maps.InfoWindow({
            					content: '<div id="content">'+
            				      '<div id="siteNotice">'+
            				      '</div>'+
            				      '<h1 id="firstHeading" class="firstHeading">'+ htmlEncode(service.name) +'</h1>'+
            				      '<div id="bodyContent">'+
            				      '<p>'+ htmlEncode(service.description) +'</p>'+
            				      '</div>'+
            				      '</div>'
            				});
            				google.maps.event.addListener(map, 'closeclick', function () {
            					infoWindow = null;
            			    });
            				infoWindow.open(map, marker);
            			});
            		}
        		});
        		
        	},
        	error: function () {
        		if (console)
        			console.log("An error occurred while loading data. Please reload your browser.");
        	}
        });
    };
    google.maps.event.addListener(map, 'bounds_changed', function () {
    	getServices(getSearchCriteria(), null);
    });
    
});

File: src/rogerthat/restapi/service_map.py
# -*- coding: utf-8 -*-
# Copyright 2019 Green Valley Belgium NV
# NOTICE: THIS FILE HAS BEEN MODIFIED BY GREEN VALLEY BELGIUM NV IN ACCORDANCE WITH THE APACHE LICENSE VERSION 2.0
# Copyright 2018 GIG Technology NV
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# @@license_version:1.6@@

from rogerthat.bizz.service import SERVICE_LOCATION_INDEX
from rogerthat.to.service_map import ServiceMapTO, ServiceLocationTO
from rogerthat.utils import email_to_id
from rogerthat.utils.crypto import sha256_hex
from google.appengine.api import search
from mcfw.restapi import rest
from mcfw.rpc import returns, arguments


@rest("/mobi/rest/service_map/load", "get")
@returns(ServiceMapTO)
@arguments(distance=int, lat=float, lon=float, app_id=unicode, cursor=unicode)
def load_service_locations(distance, lat, lon, app_id=None, cursor=None):

    def get_location_sort_options(lat, lon):
        loc_expr = "distance(location, geopoint(%f, %f))" % (lat, lon)
        sort_expr = search.SortExpression(expression=loc_expr,
                                          direction=search.SortExpression.ASCENDING,
                                          default_value=distance)
        return search.SortOptions(expressions=[sort_expr])

    index = search.Index(name=SERVICE_LOCATION_INDEX)
    query_string = (u"app_ids:%s" % app_id) if app_id else u""
    query = search.Query(query_string=query_string,
                         options=search.QueryOptions(returned_fields=['service', 'name', 'location', 'description'],
                                                     sort_options=get_location_sort_options(lat, lon),
                                                     limit=100,
                                                     cursor=search.Cursor(cursor)))
    search_result = index.search(query)
    result = ServiceMapTO()
    def map_result(service_search_result):
        service = ServiceLocationTO()
        for field in service_search_result.fields:
            if field.name == "service":
                service.id = unicode(email_to_id(field.value))
                service.hash = unicode(sha256_hex(field.value))
                continue
            elif field.name == "name":
                service.name = field.value
                continue
            elif field.name == "location":
                service.lat = field.value.latitude
                service.lon = field.value.longitude
                continue
            elif field.name == "description":
                service.description = field.value
                continue
        return service
    result.services = [map_result(r) for r in search_result.results]
    result.cursor = search_result.cursor.web_safe_string if search_result.cursor else None
    return result
